Change the bundle and ndjosn Folder path in E2E Code.

vmsuthar56 opened the pull request:
Change the bundle and ndjosn Folder path in E2E Code.

mikaelweave commented on src/FhirLoader.Common/BundleClient.cs:
```diff
@@ -49,6 +49,8 @@ public async Task Send(ProcessedBundle bundle, Action<int, long>? metricsCallbac
             var content = new StringContent(bundle.BundleText!, Encoding.UTF8, "application/json");
             HttpResponseMessage response;
 
+            var requestUri = !string.IsNullOrEmpty(bundle.BundleUri) ? $"/{bundle.ResourceType}?url={bundle.BundleUri}" : string.Empty;
```
requestUrl should be:
- `/` for posting of bundles
- `/*resourceType*` for resources that don't have an id
- `/*resourceType*/*id*` for resources that do have an id

vmsuthar56 replied:
For posting bundle, in command we have "/" at the end of FHIR Url.
In existing implementation requestUri was empty string for bundle, we have kept it as it in case of bundle.
We have not encountered a scenario where we don't have Id for resource, could you please share example.
For posting profile resources, requestUri would be "/resourceType".
Referring to https://learn.microsoft.com/en-us/azure/healthcare-apis/azure-api-for-fhir/store-profiles-in-fhir

mikaelweave replied:
- Profiles should generally always be PUT commands to `/*resourceType*/*id*
- Other "single resources" could be PUT to `/*resourceType*/*id* or POST to `/*resourceType*` if there is no ID
- Bundles are a special case and are always POST to `/`,
  - Fun fact, if you POST a bundle to `/Bundle` it will not actually be processed but stored instead.

mikaelweave commented on src/FhirLoader.Common/BundleClient.cs:
```diff
@@ -57,7 +59,7 @@ public async Task Send(ProcessedBundle bundle, Action<int, long>? metricsCallbac
 
                 response = await _resiliencyStrategy.ExecuteAsync(
-                    async ct => await _client.PostAsync("", content, ct),
+                    async ct => await _client.PostAsync(requestUri, content, ct),
```
This should be POST in most situations, unless the `requestUri` has an `id` at the end. This will allow loading the same profile multiple times against a FHIR service

vmsuthar56 replied:
This is post only.
For storing profiles, we are referring to https://learn.microsoft.com/en-us/azure/healthcare-apis/azure-api-for-fhir/store-profiles-in-fhir
This page has section: Accessing profiles and storing profiles
from this section we followed the given sample which has URL with "url" parameter like below for POST:
https://myAzureAPIforFHIR.azurehealthcareapis.com/StructureDefinition?url=http://hl7.org/fhir/us/core/StructureDefinition/us-core-allergyintolerance

Thats why we used it that way for post.
Now we have changed it to use resource type only like below.
http://whatever/StructureDefinition

mikaelweave replied:
A customer should be able to run the implementation guide loading multiple times without failure. Also the documentation is wrong because if you are sending a resource to FHIR with an `id`, it generally expected that that will be persisted. Moreover, these resources are standard, and I believe customers would expect them to be at the `id` location as defined in the profile.

FYI @EXPEkesheth - I will file a PR on the docs page for you to look at and approve/deny.

mikaelweave commented on src/FhirLoader.Common/BundleClient.cs:
```diff
@@ -87,7 +89,7 @@ public async Task Send(ProcessedBundle bundle, Action<int, long>? metricsCallbac
                 {
                     _logger.LogError(responseString);
-                }   
+                }
```
Can we rename this class `FhirResourceClient` instead of `BundleClient`?

vmsuthar56 replied:
Renamed as FhirResourceClient.

mikaelweave commented on src/FhirLoader.Common/ProcessedBundle.cs:
```diff
@@ -11,5 +11,7 @@ public class ProcessedBundle
         public string? BundleText;
         public int BundleCount;
         public string? BundleFileName;
+        public string? ResourceType;
```
Since this isn't always a bundle, can we change this to "ProcessedResource"? Or add a new class for Resource so it's clear when we are using a bundle or not?

vmsuthar56 replied:
Changed to ProcessedResource.

mikaelweave commented on src/FhirLoader.Common/ProfileFileHandler.cs:
```diff
@@ -0,0 +1,50 @@
+            using (StreamReader reader = new StreamReader(_inputStream))
+                bundle = JObject.Parse(reader.ReadToEnd());
+
+            yield return new ProcessedBundle
```
Since this is just a bundle of 1, would be best to just return an object representing the resource without the bundle wrapper. This would make the code more clear.

vmsuthar56 replied:
Do you mean to remove the processhandler and return the JObject or remove the yield? If we remove the processhandler, then we have to change the code from the program.cs file to the FhirResourceClient send method. Can you please elaborate it more?

mikaelweave replied:
Actually I misread this code - it looks good! I just thought the `BundleText` was a resource of type `Bundle` but was just confused by the property names.

mikaelweave commented on src/FhirLoader.Common/SourceFileHandler.cs:
```diff
@@ -1,7 +1,8 @@
 using Azure.Storage.Blobs.Models;
 using Microsoft.Extensions.Logging;
-
+using Microsoft.VisualBasic;
```
Don't think this using is needed?

vmsuthar56 replied:
Removed.

mikaelweave commented on src/FhirLoader.Common/SourceFileHandler.cs:
```diff
@@ -69,13 +70,58 @@ public IEnumerable<FhirFileHandler> LoadFromFilePath(string bundlePath, int bund
                 if (filePath.EndsWith(".json"))
                 {
-                    yield return new BundleFileHandler(safeFileStream, filePath, bundleSize, _logger);
+                    if (CheckResourceType(filePath)?.ToLower() == "bundle")
```
Ideally, the CLI could load single resources or bundles. May be out of scope for this story, but that's the future goal. 😄

mikaelweave reviewed (commented):
A few suggestions - it would be nice to allow both Bundles and standard FHIR resources to be  processed by this tool. That is the ideal future state. Can we make this shift as part of the PR?

mikaelweave commented on src/FhirLoader.Common/FhirResourceClient.cs:
```diff
@@ -44,20 +44,22 @@ public BundleClient(string baseUrl, ILogger logger, string? tenantId = null, Asy
             try
             {
-                _logger.LogTrace($"Sending {bundle.BundleCount} resources to {_client.BaseAddress}...");
+                _logger.LogTrace($"Sending {bundle.ResourceCount} resources to {_client.BaseAddress}...");
```
Can we add `bundle.ResourceType` to the message so users know what is being sent?

vmsuthar56 replied:
Do you mean to print ResourceType as "bundle" for ndjson and bundle and for profile will read from processedResource.ResourceType?

mikaelweave replied:
Yes! Exactly. So users know what they have sent. Super useful for profiles.

vmsuthar56 replied:
Added Resource Type to the message string.

mikaelweave commented on src/FhirLoader.Common/FhirResourceClient.cs:
```diff
@@ -44,20 +44,22 @@ public BundleClient(string baseUrl, ILogger logger, string? tenantId = null, Asy
 
                 response = await _resiliencyStrategy.ExecuteAsync(
-                    async ct => await _client.PostAsync("", content, ct),
+                    async ct => await _client.PostAsync(requestUri, content, ct),
```
This may need to be a `PUT` following the below logic:
- If the current object being sent is a bundle, `POST`
- If the current object is not a bundle, but doesn't have an `id` property in the JSON, `POST`
- If the current object is not a bundle, but *DOES* have an `id` property in the JSON, `PUT`

vmsuthar56 replied:
If I consider the scenirio for ndjson, then it works, but for the package third point, it won't work 
because in profile jsons we do have an id value (e.g us-core-client,us-core-allergyintolerance-clinical-status) without any guid. So in this case, it will always post. 
So bundle and ndjson have this implementation already, where in the handler it checks the id and defines the method post or put.

vmsuthar56 replied:
@mikaelweave Logic has been changed to handle the above cases in the FhirResourceClient.cs file. Please verify and share your thoughts.

mikaelweave replied:
Looks great @vmsuthar56!

mikaelweave commented on test/FhirLoader.Tests.E2E/FHIRLoaderE2ETest.cs:
```diff
@@ -93,5 +94,152 @@ public async void BlobFhir_ndjson_Test()
+        /// If a file without a bundle resource type exists in a bundle folder, then the tool will skip those files.
+        /// </summary>
+        [Fact]
+        public void CheckIfFileswithoutbundleresourceTypeExist_Test()
```
What is this testing exactly? I don't see any execution of the program code in this test.

vmsuthar56 replied:
Tool will skip those files which do not have bundle resource type.

mikaelweave commented on test/FhirLoader.Tests.E2E/FHIRLoaderE2ETest.cs:
```diff
@@ -93,5 +94,152 @@ public async void BlobFhir_ndjson_Test()
+        /// returns true if matched, otherwise false.
+        /// </summary>
+        [Fact]
+        public void checkIfPackageTypeExists_Test()
```
What exactly is this testing? I don't see any execution of the program code.

vmsuthar56 replied:
I have changed the logic for this case. Instead of a static value, it will read the file from the folder and check if the type in the package json does not match values such as conformance, fhir.ig, and core, then throw an exception.

mikaelweave reviewed (changes requested):
Few more comments here

vmsuthar56 commented on src/FhirLoader.Common/FhirResourceClient.cs:
```diff
@@ -44,20 +44,22 @@ public BundleClient(string baseUrl, ILogger logger, string? tenantId = null, Asy
         }
 
-        public async Task Send(ProcessedBundle bundle, Action<int, long>? metricsCallback = null, CancellationToken? cancel = null)
+        public async Task Send(ProcessedResource bundle, Action<int, long>? metricsCallback = null, CancellationToken? cancel = null)
```
Renamed as processedResource

vmsuthar56 commented on src/FhirLoader.Common/FhirResourceClient.cs:
```diff
@@ -44,20 +44,22 @@ public BundleClient(string baseUrl, ILogger logger, string? tenantId = null, Asy
+            var content = new StringContent(bundle.ResourceText!, Encoding.UTF8, "application/json");
             HttpResponseMessage response;
 
+            var requestUri = !string.IsNullOrEmpty(bundle.ResourceType) ? $"/{bundle.ResourceType}" : string.Empty;
```
added  IsBundle flag.

mikaelweave commented:
I'm getting an error trying to run the fhir-loader in a realworld scenario. Can you test this out and tell me if it works for you??

`npm --registry https://packages.simplifier.net install hl7.fhir.r4.core@4.0.1`
`./bin/Debug/net6.0/FhirLoader.Tool --package node_modules/hl7.fhir.r4.core/ --fhir https://7fzg5udn3j3d6ahds-testdata.fhir.azurehealthcareapis.com`

vmsuthar56 commented:
> 
@mikaelweave , We have resolved the issue. We didn't have the proper name for the fhir core package type, so we added Core as  type in the PackageTypeHelper.cs. We have renamed it to fhircore.

mikaelweave commented:
I'm still having an issue :( 

Will merge and look as we need this for another project

![CleanShot 2022-10-05 at 09 25 47](https://user-images.githubusercontent.com/753437/194112082-a1e6e7f1-b829-4b0c-bb97-f82181d030d1.png)

mikaelweave commented:
It didn't work initially on my Mac, but I just needed to add the full file path and we're good! I added a function to put these in bundles but found it didn't speed up processing vs the `--concurrency 50` option. Tested with the VSAC package which is like 12,000 resources and it worked GREAT!!!
